### PR TITLE
Leeky effects

### PR DESCRIPTION
**Pull request by The1Penguin**
So I got introduced to [effectful](https://hackage.haskell.org/package/effectful) and also watched [this video](https://www.youtube.com/watch?v=lRU6TDgadqw). This got me inspired to replace our mtl stack with effects, since we do not use a lot and this looked cool ;P

### CODE REVIEW

File: app/Main.hs [hunk 1]
```diff
@@ -93,25 +95,24 @@ main = (recreateConfig . getOpt Permute opts <$> getArgs) >>= \case
...
+  -> Eff es ()
 webserver Config{_cPort=webserverPort} viewRef upd =
-  run webserverPort $ foldr
+  liftIO . run webserverPort $ foldr
     (logStdout . static $(embedDir "static") <$>)
     (notFound (send $ html "not found..."))
     [ get "/"  (liftIO (readIORef viewRef) >>= send . html . encodeUtf8 . render)
-    , get "/r" (liftIO (tryPutMVar upd ()) >> send (redirect302 "/"))
+    , get "/r" (liftIO (runEff . runConcurrent $ tryPutMVar upd ()) >> send (redirect302 "/"))
     ]
```
Comment (The1Penguin) [inline app/Main.hs:120]: `webserver` looks a bit ugly, if someone wants to figure it out, be my guest

**Comment by The1Penguin** [comment]
Overall, think this shows a bit more clearly where the effects come from

File: app/Main.hs (hunk 1 above)
Comment (Rembane) [reply app/Main.hs:120]: I gave it a go and couldn't make it less ugly. What makes it ugly btw? :joy_cat:

Comment (The1Penguin) [reply app/Main.hs:120]: It's mostly that warp's `run` is of type `Port -> Application -> IO ()`, and that `(liftIO (runEff . runConcurrent $ tryPutMVar upd ()) >> send (redirect302 "/"))` has type `ResponderM`. This means that we have to run `runEff . runConcurrent $ tryPutMVar upd ()` to access the MVar from Effectful.

One potential fix is to extend warp and twain, but that seems like a lot of work

Comment (Rembane) [reply app/Main.hs:120]: Yeah, that's not super beautiful. It's like a cake with `liftIO` sandwiched between almost every layer. I might have another variant of this code for you, let me put it in a commit. 

Extending warp and twain would be exciting but perhaps a bit too ambitious, especially in this PR. :smile_cat: 

Another ambitious route is to use Servant instead and make things even more typed! Here's the box of Pandora: https://docs.servant.dev/en/stable/cookbook/hoist-server-with-context/HoistServerWithContext.html

Comment (Rembane) [reply app/Main.hs:120]: After some experimenting back and forth I came to the conclusion that I do not have another variant. :smile:

**Review by Rembane — APPROVED** [review]
Looking good. I've heard good things about leeky effects. Lets see if they can deliver food too!

**Pull request merged**